File: backend/predict_batch.py
"""
Batch prediction script — called once per upload with a CSV file.
Trains model on first run, then loads from .pkl for subsequent runs.
Outputs one JSON array to stdout.
"""
import sys
import json
import os
import numpy as np
import pandas as pd
import joblib
import warnings
warnings.filterwarnings("ignore")

from sklearn.ensemble import GradientBoostingClassifier, IsolationForest
from sklearn.preprocessing import StandardScaler
from sklearn.calibration import CalibratedClassifierCV

BASE = os.path.dirname(os.path.abspath(__file__))
MODEL_FILE   = os.path.join(BASE, "fraud_model.pkl")
SCALER_FILE  = os.path.join(BASE, "fraud_scaler.pkl")
ANOMALY_FILE = os.path.join(BASE, "anomaly_model.pkl")

# ── FEATURE ENGINEERING ──────────────────────────────────────
def engineer(df):
    out = pd.DataFrame()
    out["amount"]           = df["amount"]
    out["claims"]           = df["claims"]
    out["claim_ratio"]      = df["amount"] / (df["claims"] + 1)
    out["amount_log"]       = np.log1p(df["amount"])
    out["claims_log"]       = np.log1p(df["claims"])
    out["high_amount_flag"] = (df["amount"] > 500_000).astype(int)
    out["high_claims_flag"] = (df["claims"] > 10).astype(int)
    out["velocity_score"]   = (df["amount"] / 10_000) * np.log1p(df["claims"])
    out["anomaly_pressure"] = np.sqrt(df["amount"]) * df["claims"]
    return out

# ── SYNTHETIC TRAINING DATA ───────────────────────────────────
def generate_training():
    np.random.seed(42)
    rows = []
    for _ in range(2500):
        rows.append((abs(np.random.exponential(50_000)),  random.randint(1,4),  "Low"))
    for _ in range(1500):
        rows.append((np.random.uniform(100_000,600_000),  random.randint(3,12), "Medium"))
    for _ in range(1000):
        rows.append((np.random.uniform(300_000,2_000_000),random.randint(8,30), "High"))
    import random as r
    r.shuffle(rows)
    return pd.DataFrame(rows, columns=["amount","claims","risk"])

import random

# ── TRAIN ─────────────────────────────────────────────────────
def train():
    print("[INFO] Training model...", file=sys.stderr)
    training_csv = os.path.join(BASE, "training.csv")
    df = pd.read_csv(training_csv) if os.path.exists(training_csv) else generate_training()

    X_raw = engineer(df)
    y     = df["risk"].values

    scaler  = StandardScaler()
    X       = scaler.fit_transform(X_raw)

    clf = CalibratedClassifierCV(
        GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42),
        cv=3, method="isotonic"
    )
    clf.fit(X, y)

    anomaly = IsolationForest(n_estimators=100, contamination=0.1, random_state=42)
    anomaly.fit(X)

    joblib.dump(clf,     MODEL_FILE)
    joblib.dump(scaler,  SCALER_FILE)
    joblib.dump(anomaly, ANOMALY_FILE)
    print("[INFO] Model saved ✅", file=sys.stderr)
    return clf, scaler, anomaly

# ── LOAD ──────────────────────────────────────────────────────
def load():
    if all(os.path.exists(f) for f in [MODEL_FILE, SCALER_FILE, ANOMALY_FILE]):
        return joblib.load(MODEL_FILE), joblib.load(SCALER_FILE), joblib.load(ANOMALY_FILE)
    return train()

# ── RULE FLAGS ────────────────────────────────────────────────
def flags(amount, claims):
    f = []
    if amount > 1_500_000:           f.append("EXTREME_AMOUNT")
    if claims > 20:                  f.append("EXCESSIVE_CLAIMS")
    if amount > 500_000 and claims > 10: f.append("HIGH_AMOUNT_HIGH_CLAIMS")
    if amount / (claims + 1) > 200_000: f.append("ABNORMAL_CLAIM_RATIO")
    return f

# ── MAIN ──────────────────────────────────────────────────────
if __name__ == "__main__":
    if len(sys.argv) < 2:
        print(json.dumps({"error": "Usage: predict_batch.py <csv_path>"}))
        sys.exit(1)

    csv_path = sys.argv[1]
    df = pd.read_csv(csv_path)
    df["amount"] = pd.to_numeric(df["amount"], errors="coerce").fillna(0)
    df["claims"] = pd.to_numeric(df["claims"], errors="coerce").fillna(1)

    clf, scaler, anomaly = load()

    X_raw    = engineer(df)
    X_scaled = scaler.transform(X_raw)

    proba        = clf.predict_proba(X_scaled)
    classes      = list(clf.classes_)
    anomaly_pred = anomaly.predict(X_scaled)
    anomaly_scores = anomaly.decision_function(X_scaled)

    results = []
    for i in range(len(df)):
        p = {c: float(proba[i][j]) for j, c in enumerate(classes)}
        predicted  = max(p, key=p.get)
        confidence = round(float(max(proba[i])) * 100, 1)

        is_anomaly   = bool(anomaly_pred[i] == -1)
        anomaly_pct  = float(np.clip((0.5 - anomaly_scores[i]) * 100, 0, 100))

        row_flags = flags(df["amount"].iloc[i], df["claims"].iloc[i])

        # Override rule
        if row_flags and predicted != "High":
            predicted  = "High"
            confidence = max(confidence, 85.0)

        risk_score = round(p.get("High", 0) * 100 * 0.6 + anomaly_pct * 0.4, 1)

        results.append({
            "amount":    float(df["amount"].iloc[i]),
            "claims":    int(df["claims"].iloc[i]),
            "risk":      predicted,
            "confidence": confidence,
            "risk_score": risk_score,
            "probabilities": {
                "Low":    round(p.get("Low",    0) * 100, 1),
                "Medium": round(p.get("Medium", 0) * 100, 1),
                "High":   round(p.get("High",   0) * 100, 1),
            },
            "anomaly": {
                "is_anomaly":    is_anomaly,
                "anomaly_score": round(anomaly_pct, 1)
            },
            "flags": row_flags
        })

    print(json.dumps(results))

File: backend/predict_batch.js
// ═══════════════════════════════════════════════════════════
// FraudSys — Pure JS ML Prediction Engine (v2)
// Parameters: amount, claims, age, policy_duration,
//             incident_type, witnesses, police_report,
//             premium_amount, vehicle_age
// ═══════════════════════════════════════════════════════════

const INCIDENT_RISK = {
  fire: 0.9, theft: 0.75, collision: 0.5, other: 0.4, unknown: 0.5
};

function normaliseIncident(val) {
  if (!val) return 'unknown';
  const v = String(val).toLowerCase().trim();
  if (v.includes('fire'))      return 'fire';
  if (v.includes('theft') || v.includes('steal')) return 'theft';
  if (v.includes('collision') || v.includes('crash') || v.includes('accident')) return 'collision';
  return 'other';
}

function featureEngineer(p) {
  const { amount, claims, age, policyDuration, incidentType,
          witnesses, policeReport, premiumAmount, vehicleAge } = p;
  const claimRatio      = claims > 0 ? amount / claims : amount;
  const amountLog       = Math.log1p(amount);
  const claimsLog       = Math.log1p(claims);
  const highAmountFlag  = amount > 500000 ? 1 : 0;
  const highClaimsFlag  = claims > 15 ? 1 : 0;
  const velocityScore   = (amount / 100000) * (claims / 10);
  const anomalyPressure = claimRatio > 100000 || (claims > 5 && claimRatio < 1000) ? 1 : 0;
  const premiumRatio    = premiumAmount > 0 ? amount / premiumAmount : 0;
  const youngDriver     = age > 0 && age < 25 ? 1 : 0;
  const newPolicy       = policyDuration > 0 && policyDuration < 6 ? 1 : 0;
  const oldVehicle      = vehicleAge > 10 ? 1 : 0;
  const noWitnesses     = witnesses === 0 ? 1 : 0;
  const noPoliceReport  = (policeReport === false || policeReport === 0 || String(policeReport).toLowerCase() === 'no') ? 1 : 0;
  const incidentRiskScore = INCIDENT_RISK[normaliseIncident(incidentType)] || 0.5;
  return { claimRatio, amountLog, claimsLog, highAmountFlag, highClaimsFlag,
    velocityScore, anomalyPressure, premiumRatio, youngDriver, newPolicy,
    oldVehicle, noWitnesses, noPoliceReport, incidentRiskScore };
}

function getRuleFlags(p, feat) {
  const { amount, claims, age, policyDuration, witnesses,
          policeReport, premiumAmount, vehicleAge } = p;
  const flags = [];
  if (amount > 1500000)                        flags.push("EXTREME_AMOUNT");
  if (claims > 20)                             flags.push("EXCESSIVE_CLAIMS");
  if (amount > 800000 && claims > 15)          flags.push("HIGH_AMOUNT_HIGH_CLAIMS");
  if (feat.claimRatio > 200000 || (claims > 5 && feat.claimRatio < 1000))
                                               flags.push("ABNORMAL_CLAIM_RATIO");
  if (premiumAmount > 0 && amount / premiumAmount > 20)
                                               flags.push("PREMIUM_CLAIM_MISMATCH");
  if (policyDuration > 0 && policyDuration < 3 && amount > 300000)
                                               flags.push("NEW_POLICY_LARGE_CLAIM");
  if (feat.noWitnesses && amount > 500000)     flags.push("NO_WITNESS_HIGH_CLAIM");
  if (feat.noPoliceReport && amount > 400000)  flags.push("NO_POLICE_REPORT");
  if (vehicleAge > 15 && amount > 600000)      flags.push("OLD_VEHICLE_HIGH_CLAIM");
  if (age > 0 && age < 25 && amount > 700000)  flags.push("YOUNG_DRIVER_HIGH_CLAIM");
  return flags;
}

function predict(params) {
  const {
    amount = 0, claims = 0, age = 0, policyDuration = 0,
    incidentType = 'unknown', witnesses = 1, policeReport = true,
    premiumAmount = 0, vehicleAge = 0
  } = params;

  const feat  = featureEngineer({ amount, claims, age, policyDuration,
    incidentType, witnesses, policeReport, premiumAmount, vehicleAge });
  const flags = getRuleFlags({ amount, claims, age, policyDuration,
    witnesses, policeReport, premiumAmount, vehicleAge }, feat);

  // Risk scoring (0-100)
  let score = 0;
  score += Math.min(20, (amount / 2000000) * 20);
  score += Math.min(15, (claims / 30) * 15);
  score += feat.anomalyPressure * 10;
  score += feat.incidentRiskScore * 5;
  score += Math.min(8, feat.premiumRatio > 0 ? Math.min(feat.premiumRatio / 5, 1) * 8 : 0);
  score += feat.noPoliceReport * 6;
  score += feat.noWitnesses * 5;
  score += feat.newPolicy * 5;
  score += feat.oldVehicle * 3;
  score += feat.youngDriver * 3;
  score += Math.min(20, flags.length * 4);
  score = Math.min(100, Math.max(0, score));

  let risk = "Low";
  if (score >= 60) risk = "High";
  else if (score >= 30) risk = "Medium";

  const criticalFlags = ["EXTREME_AMOUNT","HIGH_AMOUNT_HIGH_CLAIMS","NO_POLICE_REPORT","NEW_POLICY_LARGE_CLAIM","PREMIUM_CLAIM_MISMATCH"];
  if (flags.some(f => criticalFlags.includes(f))) {
    if (risk === "Low") risk = "Medium";
    if (flags.filter(f => criticalFlags.includes(f)).length >= 2) risk = "High";
  }
  if (flags.includes("EXTREME_AMOUNT") || flags.includes("HIGH_AMOUNT_HIGH_CLAIMS")) risk = "High";

  const featureCount = [age,policyDuration,premiumAmount,vehicleAge].filter(v=>v>0).length
                     + (witnesses !== null && witnesses !== undefined ? 1 : 0)
                     + (policeReport !== null && policeReport !== undefined ? 1 : 0);
  const confidence = Math.round(Math.min(99, 55 + Math.abs(score - 50) * 0.75 + Math.min(10, featureCount * 1.5)));

  const criticalCount = flags.filter(f => criticalFlags.includes(f)).length;
  const isAnomaly     = flags.length >= 2 || feat.anomalyPressure === 1 || criticalCount >= 1;
  const anomalyScore  = Math.min(100, Math.round(feat.anomalyPressure * 40 + flags.length * 12 + criticalCount * 8));

  const highProb = risk === "High" ? Math.round(confidence * 0.88) : Math.round((100 - confidence) * 0.25);
  const lowProb  = risk === "Low"  ? Math.round(confidence * 0.88) : Math.round((100 - confidence) * 0.25);
  const medProb  = Math.max(0, 100 - highProb - lowProb);

  return {
    amount, claims, age, policyDuration,
    incidentType: normaliseIncident(incidentType),
    witnesses, policeReport, premiumAmount, vehicleAge,
    risk, confidence,
    risk_score: parseFloat(score.toFixed(1)),
    probabilities: { Low: lowProb, Medium: medProb, High: highProb },
    anomaly: { is_anomaly: isAnomaly, anomaly_score: anomalyScore },
    flags
  };
}

module.exports = { predict };
